File: project/src/analysis.js
import * as math from 'mathjs';
import { RandomForestClassifier } from 'ml-random-forest';

export function calculateHRMetrics(df) {
  const attritionRate = (df['Attrition'].sum() / df.length) * 100;
  const avgSatisfaction = math.mean(df['JobSatisfaction']);
  const avgYearsCompany = math.mean(df['YearsAtCompany']);
  const avgMonthlyIncome = math.mean(df['MonthlyIncome']);

  return {
    attrition_rate: attritionRate,
    avg_satisfaction: avgSatisfaction,
    avg_years_company: avgYearsCompany,
    avg_monthly_income: avgMonthlyIncome
  };
}

export function departmentAnalysis(df) {
  const departments = [...new Set(df['Department'])];
  return departments.map(dept => {
    const deptData = df.filter(row => row.Department === dept);
    return {
      Department: dept,
      Attrition: math.mean(deptData['Attrition']),
      JobSatisfaction: math.mean(deptData['JobSatisfaction']),
      MonthlyIncome: math.mean(deptData['MonthlyIncome']),
      YearsAtCompany: math.mean(deptData['YearsAtCompany'])
    };
  });
}

export class AttritionPredictor {
  constructor() {
    this.model = new RandomForestClassifier({
      nEstimators: 100,
      maxDepth: 10,
      seed: 42
    });
  }

  train(X, y) {
    const trainSize = Math.floor(X.length * 0.8);
    const X_train = X.slice(0, trainSize);
    const X_test = X.slice(trainSize);
    const y_train = y.slice(0, trainSize);
    const y_test = y.slice(trainSize);

    this.model.train(X_train, y_train);
    const y_pred = this.model.predict(X_test);

    return {
      accuracy: this.calculateAccuracy(y_test, y_pred),
      confusion_matrix: this.calculateConfusionMatrix(y_test, y_pred)
    };
  }

  calculateAccuracy(actual, predicted) {
    const correct = actual.filter((val, idx) => val === predicted[idx]).length;
    return correct / actual.length;
  }

  calculateConfusionMatrix(actual, predicted) {
    const matrix = [[0, 0], [0, 0]];
    actual.forEach((val, idx) => {
      matrix[val][predicted[idx]]++;
    });
    return matrix;
  }
}

File: project/src/analysis.py
import pandas as pd
import numpy as np
from typing import Dict, List
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix

def calculate_hr_metrics(df: pd.DataFrame) -> Dict[str, float]:
    """
    Calculate key HR metrics
    """
    metrics = {
        'attrition_rate': (df['Attrition'].sum() / len(df)) * 100,
        'avg_satisfaction': df['JobSatisfaction'].mean(),
        'avg_years_company': df['YearsAtCompany'].mean(),
        'avg_monthly_income': df['MonthlyIncome'].mean()
    }
    return metrics

def department_analysis(df: pd.DataFrame) -> pd.DataFrame:
    """
    Analyze metrics by department
    """
    dept_metrics = df.groupby('Department').agg({
        'Attrition': 'mean',
        'JobSatisfaction': 'mean',
        'MonthlyIncome': 'mean',
        'YearsAtCompany': 'mean'
    }).round(2)
    
    return dept_metrics

class AttritionPredictor:
    def __init__(self, random_state: int = 42):
        self.model = RandomForestClassifier(random_state=random_state)
        
    def train(self, X: pd.DataFrame, y: pd.Series) -> Dict:
        """
        Train the attrition prediction model
        """
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=0.2, random_state=42
        )
        
        self.model.fit(X_train, y_train)
        y_pred = self.model.predict(X_test)
        
        return {
            'classification_report': classification_report(y_test, y_pred, output_dict=True),
            'confusion_matrix': confusion_matrix(y_test, y_pred).tolist()
        }
    
    def get_feature_importance(self, feature_names: List[str]) -> Dict[str, float]:
        """
        Get the importance of each feature in predicting attrition
        """
        importance = self.model.feature_importances_
        return dict(zip(feature_names, importance))